### DISCUSSION TITLE
Sourcing data from S3 with targets

### DISCUSSION THREAD

**Discussion by hershelm** [ropensci/targets#872] [Help] [answered]
Hi there, I'm really enjoying this package, thanks so much for your work.

I'm attempting to create a targets workflow for the drake workflow described here: https://mdneuzerling.com/post/sourcing-data-from-s3-with-drake/

I am struggling to get targets to recognize a change in an object's etag. Specifically, I'm not sure how to compare to the previous file's etag. 

I get the following

``` r
library(targets)
library(tidyverse)
library(tarchetypes)
library(aws.s3)
library(paws)


object_name <- "xxx"
bucket_name <- "xxx"


get_etag <- function(bucket, object) {
  
  client <- paws::s3()
  
  etag <- 
    client$get_object(Bucket = bucket,
                      Key = object) %>% 
    .$ETag
  
  etag
}


get_data <- function(bucket, object, etag) {
  
  aws.s3::s3read_using(arrow::read_parquet,
                       object = object_name,
                       bucket = bucket_name)
}


tar_plan(
  tar_target(
    test,
    get_data(
      object = object_name,
    ),
    cue = tar_force_change(condition = get_etag(bucket = bucket_name,
                                                object = object_name))
    
  )
  
)
#> Warning in any(condition): coercing argument of type 'character' to logical
#> Error in if (any(condition)) {: missing value where TRUE/FALSE needed
```

<sup>Created on 2022-07-14 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by hershelm** [comment]
I understand the error message, but am wondering -- in `drake` it appeared it tracked the etag, and sensed when it changed. Is there an equivalent in `targets`?

What is the best practice for this workflow of detecting a change with a previous etag?

**Comment by hershelm** [comment, marked as answer]
After some testing, I found that `tarchetypes::tar_change` works great

```
tar_plan(
  tar_change(
    test,
    get_data(
      bucket = bucket_name,
      object = object_name
    ),
    change = get_etag(bucket = bucket_name,
                      object = object_name)
  )
)
```

leaving here, in case anyone comes across something similar

**Reply by philiporlando** [reply]
Thanks for sharing this! Any thoughts on how to incorporate this with a target that returns a vector containing multiple objects?

I have a project with thousands of objects within a given bucket/prefix, and I'm trying to implement a branching pattern to simplify my target definitions.

**Reply by wlandau** [reply]
Maybe `tar_target(format = "url")` or  `tarchetypes::tar_url()` if you can reference the S3 object as a URL? `format = "url"` checks the ETag and Last-Modified metadata for changes.

**Reply by philiporlando** [reply]
This sounds promising. Thanks for the suggestion!
